src/storage/v2/storage.cpp

```diff
@@ -498,1825 +60,45 @@ Storage::Accessor::Accessor(Storage *storage, IsolationLevel isolation_level, St
-                    [&](const auto &other_client) { return client->Name() == other_client->Name(); })) {
-      return RegisterReplicaError::NAME_EXISTS;
-    }
-
-    if (std::any_of(clients.begin(), clients.end(),
-                    [&client](const auto &other_client) { return client->Endpoint() == other_client->Endpoint(); })) {
-      return RegisterReplicaError::END_POINT_EXISTS;
-    }
-
-    clients.push_back(std::move(client));
-    return {};
-  });
-  spdlog::info("Replica {} registered.", name);
-}
-
-bool Storage::UnregisterReplica(const std::string &name) {
-
-  spdlog::trace("Unregistering replica...");
-  MG_ASSERT(replication_role_.load() == replication::ReplicationRole::MAIN,
-            "Only main instance can unregister a replica!");
-  if (ShouldStoreAndRestoreReplicationState()) {
-    if (!storage_->Delete(name)) {
-      spdlog::error("Error when removing replica {} from settings.", name);
-      return false;
-    }
+/// Main lock is taken by the caller.
+void Storage::SetStorageMode(StorageMode storage_mode) {
+  std::unique_lock main_guard{main_lock_};
+  MG_ASSERT(
+      (storage_mode_ == StorageMode::IN_MEMORY_ANALYTICAL || storage_mode_ == StorageMode::IN_MEMORY_TRANSACTIONAL) &&
+      (storage_mode == StorageMode::IN_MEMORY_ANALYTICAL || storage_mode == StorageMode::IN_MEMORY_TRANSACTIONAL));
+  if (storage_mode_ != storage_mode) {
+    storage_mode_ = storage_mode;
+    FreeMemory(std::move(main_guard));
   }
-
-  return replication_clients_.WithLock([&](auto &clients) {
-    return std::erase_if(clients, [&](const auto &client) { return client->Name() == name; });
-  });
 }
 
-std::optional<replication::ReplicaState> Storage::GetReplicaState(const std::string_view name) {
-  spdlog::trace("Getting replica state...");
-  return replication_clients_.WithLock([&](auto &clients) -> std::optional<replication::ReplicaState> {
-    const auto client_it =
-        std::find_if(clients.cbegin(), clients.cend(), [name](auto &client) { return client->Name() == name; });
-    if (client_it == clients.cend()) {
-      return std::nullopt;
-    }
-    return (*client_it)->State();
-  });
-}
-
-replication::ReplicationRole Storage::GetReplicationRole() const { return replication_role_; }
+IsolationLevel Storage::GetIsolationLevel() const noexcept { return isolation_level_; }
 
-std::vector<Storage::ReplicaInfo> Storage::ReplicasInfo() {
-  spdlog::trace("Getting replicas info...");
-  return replication_clients_.WithLock([](auto &clients) {
-    std::vector<Storage::ReplicaInfo> replica_info;
-    replica_info.reserve(clients.size());
-    std::transform(
-        clients.begin(), clients.end(), std::back_inserter(replica_info), [](const auto &client) -> ReplicaInfo {
-          return {client->Name(), client->Mode(), client->Endpoint(), client->State(), client->GetTimestampInfo()};
-        });
-    return replica_info;
-  });
-}
+StorageMode Storage::GetStorageMode() const { return storage_mode_; }
```

Comment:
Is `storage_mode_` protected?
Seems like we are not using the accessor when reading it. Can this cause a race condition?
*similar story for the `isolation_level_`

Reply:
for both operations you exclusive lock (main_lock_)

Reply:
You take the `main_lock_` on write/set but on on read/get.
This is why I was asking about the `Accessor`. There we take a shared lock on `main_lock_`.
We shouldn't use the `Storage::GetStorageMode()` directly since it is not protected. We should access it through the `Accessor` or add a shared_lock before. 
After our discussion on slack, I think we should add `GetStorageMode` and `GetIsolationLevel` to the `Accessor` and `DbAccessor` and replace all `interpreter_context->db->GetStorageMode()` calls with a call through an accessor.

Reply:
why not taking shared_lock in storage class then, why through accessor? Maybe I am missing something but I find it stupid to duplicate all functionalites to accessor level. Especially when main_lock_ is part of the storage

Reply:
We could take the shared_lock, but that is also currently missing from the get functions. I think...
Also, from the comments on slack, I think we are trying to limit access to the database from the queries to the appropriate accessors and not directly.